src/monodroid/jni/strings.hh

```diff
@@ -0,0 +1,772 @@
+				return false;
+			}
+
+			if (length () < s_length) {
+				return false;
+			}
+
+			return memcmp (start (), s, s_length) == 0;
+		}
+
+		template<size_t Size>
+		force_inline bool starts_with (const char (&s)[Size]) const noexcept
+		{
+			return starts_with (s, Size);
+		}
+
+		force_inline bool has_at (const char ch, size_t index) const noexcept
+		{
+			if (!can_access (index)) {
+				return false;
+			}
+
+			return start ()[index] == ch;
+		}
+
+		force_inline ssize_t find (const char ch, size_t start_index) const noexcept
+		{
+			if (!can_access (start_index)) {
+				return -1;
+			}
+
+			while (start_index <= length ()) {
+				if (start ()[start_index] == ch) {
+					return static_cast<ssize_t>(start_index);
+				}
+				start_index++;
+			}
+
+			return -1;
+		}
+
+		template<typename T>
+		force_inline bool to_integer (T &val, size_t start_index = 0, int base = 10) const noexcept
+		{
+			static_assert (std::is_integral_v<T>);
+			constexpr T min = std::numeric_limits<T>::min ();
+			constexpr T max = std::numeric_limits<T>::max ();
+			using integer = typename std::conditional<std::is_signed_v<T>, int64_t, uint64_t>::type;
+
+			if (length () == 0) {
+				return false;
+			}
+
+			if (!can_access (start_index)) {
+				log_error (LOG_DEFAULT, "Cannot convert string to integer, index %u is out of range", start_index);
+				return false;
+			}
+
+			// FIXME: this is less than ideal... we shouldn't need another buffer here
+			size_t slen = length () - start_index;
+			char s[slen + 1];
+
+			memcpy (s, start () + start_index, slen);
+			s[slen] = '\0';
+
+			integer ret;
+			char *endp;
+			bool out_of_range;
+			errno = 0;
+			if constexpr (std::is_signed_v<T>) {
+				ret = strtoll (s, &endp, base);
+				out_of_range = ret < min || ret > max;
+			} else {
+				ret = strtoull (s, &endp, base);
+				out_of_range = ret > max;
+			}
+
+			if (out_of_range || errno == ERANGE) {
+				log_error (LOG_DEFAULT, "Value %s is out of range of this type (%lld..%llu)", s, static_cast<int64_t>(min), static_cast<uint64_t>(max));
+				return false;
+			}
+
+			if (endp == s) {
+				log_error (LOG_DEFAULT, "Value %s does not represent a base %d integer", s, base);
+				return false;
+			}
+
+			if (*endp != '\0') {
+				log_error (LOG_DEFAULT, "Value %s has non-numeric characters at the end", s);
+				return false;
+			}
+
+			val = static_cast<T>(ret);
+			return true;
+		}
+
+	private:
+		force_inline bool can_access (size_t index) const noexcept
+		{
+			if (XA_UNLIKELY (!initialized () || start () == nullptr)) {
+				return false;
+			}
+
+			if (index > length ())
+				return false;
+
+			return true;
+		}
+
+	protected:
+		size_t       _last_index = 0;
+		bool         _fresh = true;
+		const char  *_start = nullptr;
+		size_t       _length = 0;
+
+		template<size_t MaxStackSize, typename TStorage, typename TChar> friend class string_base;
+	};
+
+	template<size_t MaxStackSize, bool HasResize, typename T = char>
```

Comment:
Ditto earlier comment about "premature generalization": why have `T` when we only support `char`s?

Reply:
The explanation is the same :)